### PR TITLE
4910 paging child factory

### PR DESCRIPTION
**Pull request by esaunders**
Implementation of a BaseChildFactory that can provide paging and filtering support.
This cannot be tested until other changes are made but it can be code reviewed.

### CODE REVIEW

File: Core/src/org/sleuthkit/autopsy/datamodel/BaseChildFactory.java [hunk 1]
```diff
@@ -0,0 +1,234 @@
...
+    private final Predicate<T> filter;
+    private boolean isPageChangeEvent;
+
+    private final PagingSupport pagingSupport;
+
+    /**
+     * This static map is used to facilitate communication between the UI and
+     * the child factory.
+     */
+    public static Map<String, EventBus> nodeNameToEventBusMap = new ConcurrentHashMap<>();
```
Comment (rcordovano) [inline Core/src/org/sleuthkit/autopsy/datamodel/BaseChildFactory.java]: I don't like the idea of a global variable here. Would it make sense to have static register, post, etc. methods that take a node name and delegate to the corresponding method on the fully encapsulated bus objects? Also, I am also concerned about the way update and retrieval operations can overlap in ConcurrentHashMap.

File: Core/src/org/sleuthkit/autopsy/datamodel/KnownAndSlackFilter.java [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+    @Override
+    public boolean test(T t) {
+        AbstractFile af = null;
+
+        if (t instanceof BlackboardArtifact) {
+            try {
+                af = ((BlackboardArtifact) (t)).getSleuthkitCase().getAbstractFileById(((BlackboardArtifact) t).getObjectID());
+            } catch (TskCoreException ex) {
+                Exceptions.printStackTrace(ex);
```
[the remark this answers is not in the document]
Comment (esaunders) [reply Core/src/org/sleuthkit/autopsy/datamodel/KnownAndSlackFilter.java]: This code was a temporary placeholder and no longer exists. Perhaps I should close out the older pull requests?

File: Core/src/org/sleuthkit/autopsy/datamodel/BaseChildFactory.java [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+        public int getPageCount() {
+            return pageCount;
+        }
+    }
+
+    /**
+     * Event used to let subscribers know that paging is no
+     * longer required.
+     */
+    public static class PagingDestroyedEvent {
```
[the remark this answers is not in the document]
Comment (esaunders) [reply Core/src/org/sleuthkit/autopsy/datamodel/BaseChildFactory.java]: This no longer exists.

File: Core/src/org/sleuthkit/autopsy/datamodel/BaseChildFactory.java [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+    }
+
+    /**
+     * Class that supplies paging related functionality to the base child
+     * factory class.
+     */
+    class PagingSupport {
```
[the remark this answers is not in the document]
Comment (esaunders) [reply Core/src/org/sleuthkit/autopsy/datamodel/BaseChildFactory.java]: I've made it private in the latest commit.

File: Core/src/org/sleuthkit/autopsy/datamodel/BaseChildFactory.java (hunk 1 above)
Comment (esaunders) [reply Core/src/org/sleuthkit/autopsy/datamodel/BaseChildFactory.java]: I like the idea of static methods and will look into making that change.
I don't understand your concerns about map update and retrieval. Entries are added to the map once (and never updated) when the child factory is created and they are retrieved when publishers and subscribers need them. We should probably talk about this.

Comment (rcordovano) [reply Core/src/org/sleuthkit/autopsy/datamodel/BaseChildFactory.java]: We are going with the static methods, then, coming in a later commit. I am going to take this now.

**Pull request merged**